### PR TITLE
feat(mcp): detect OAuth token boundary risks

### PR DESCRIPTION
**Pull request by SetraTheXX**
## Summary

Implements [#104](https://github.com/asamassekou10/ship-safe/issues/104) by adding four targeted OAuth security checks to `MCPSecurityAgent`:

- `MCP_UPSTREAM_TOKEN_PASSTHROUGH` — inbound MCP authorization tokens forwarded downstream
- `MCP_TOKEN_AUDIENCE_UNVALIDATED` — inbound tokens accepted without audience validation or token exchange
- `MCP_STATIC_CLIENT_ID_DYNAMIC_REG` — static OAuth client IDs combined with dynamic client registration
- `MCP_OAUTH_NO_PKCE` — authorization-code flows without effective PKCE protection

## Detection scope

The implementation is intentionally conservative.

JavaScript/TypeScript passthrough detection supports direct `.tool(...)`, `registerTool(...)`, and `addTool(...)` callback handlers, including schema/configuration arguments. Callback ownership is tied to the handler’s own function scope, so neighboring sibling functions are not incorrectly treated as MCP handlers.

Python support covers synchronous and asynchronous functions directly decorated with FastMCP `.tool(...)`, including multiline decorators. Generic `@tool` decorators are not treated as MCP handlers.

Ruby and Go passthrough detection is intentionally not claimed because reliable handler binding would require broader symbol or call-graph analysis.

Audience and PKCE source checks require an MCP-owned scope instead of relying only on a file-level MCP marker. Unrelated OAuth flows in MCP-containing files are therefore ignored. `.mcp.json` configuration scanning remains separate and unchanged.

Static client IDs and dynamic registration must belong to the same object or configuration scope. Audience validation and token exchange must be conservatively associated with the same inbound token before suppressing a finding.

Comments and non-semantic string contents are excluded from structural and mitigation matching, while required quoted object/hash keys remain supported across the validated language patterns.

`MCP_TOKEN_FORWARD_ENV` remains a separate finding. It covers credentials passed through MCP or agent configuration, while these rules focus on inbound OAuth token handling and OAuth protocol configuration.

No new dependencies, AST framework, call-graph layer, or symbol-resolution system was added.

## Validation

- Focused OAuth tests: **67/67 passed**
- Related MCP/language tests: **70/71 passed** — 1 known Windows symlink `EPERM`
- Full suite: **572/577 passed** — 5 known baseline failures
- Benchmarks: **12/12 detection scenarios**, **12/12 clean controls**
- ESLint: **0 errors**
- `git diff --check`: **clean**
- No new failures compared with the base branch

The five full-suite failures are the existing stalled-stdout and Windows symlink environment failures also present on the base branch.

Base comparison: [2bc9fe3](https://github.com/asamassekou10/ship-safe/commit/2bc9fe37295e08d601eb87673e000b494de8d90a)

Related: [#104](https://github.com/asamassekou10/ship-safe/issues/104)

### CODE REVIEW

File: cli/agents/mcp-security-agent.js [hunk 1]
```diff
@@ -249,6 +249,431 @@ const OFFICIAL_MCP_SERVERS = new Set([
...
+    .filter((scope) => (
+      scope.start <= firstIndex
+      && firstIndex < scope.end
+      && scope.start <= secondIndex
+      && secondIndex < scope.end
+      && braceDepthAt(code, scope, firstIndex) === 1
+      && braceDepthAt(code, scope, secondIndex) === 1
+    ))
+    .sort((left, right) => (left.end - left.start) - (right.end - right.start))[0] || null;
+}
```
[the remark this answers is not in the document]
Comment (SetraTheXX) [reply cli/agents/mcp-security-agent.js]: Confirmed and fixed in 4bf0aed. Added regression coverage for `{` and `}` inside string literals, and object brace depth is now calculated from the string-masked structural view while preserving source indexes and object boundaries. The suggested autofix is outdated.

File: cli/agents/mcp-security-agent.js [hunk 2]
```diff
@@ -249,6 +249,431 @@ const OFFICIAL_MCP_SERVERS = new Set([
...
+
+  const tokenReference = new RegExp(`\\b${tokenName.replace(/[.*+?^${}()|[\\]\\]/g, '\\$&')}\\b`);
+  return allRegexMatches(mitigation, scopeContent).some((match) => {
+    if (tokenReference.test(match[0])) return true;
+
+    const callEnd = scopeContent.indexOf(')', match.index);
+    const end = callEnd === -1 ? Math.min(scopeContent.length, match.index + 280) : callEnd + 1;
+    return tokenReference.test(scopeContent.slice(match.index, end));
+  });
+}
```
[the remark this answers is not in the document]
Comment (SetraTheXX) [reply cli/agents/mcp-security-agent.js]: Confirmed and fixed in 4bf0aed. Added regressions for unrelated audience validation and token exchange, plus controls for mitigations applied to the direct Authorization value. Direct candidates now require a matching masked inbound credential reference before suppressing the finding. The suggested autofix is outdated.

File: cli/agents/mcp-security-agent.js [hunk 3]
```diff
@@ -249,6 +249,437 @@ const OFFICIAL_MCP_SERVERS = new Set([
...
+        index += 1;
+        state = 'code';
+      } else {
+        blank(index);
+      }
+      continue;
+    }
+
+    if (state === 'string') {
+      if (char === '\\') {
```
[the remark this answers is not in the document]
Comment (SetraTheXX) [reply cli/agents/mcp-security-agent.js]: Confirmed and fixed in 169c991. Added regression coverage for Go raw strings containing //, /* */, #, braces, and a backslash immediately before a valid closing backtick. Go raw strings are now handled without backslash-escape semantics, while JavaScript template literals and ordinary quoted strings remain unchanged. Ruby backtick parsing and hash-rocket PKCE keys were covered in the same fix. Focused OAuth tests pass 50/50. The suggested autofix is outdated.

**Comment by asamassekou10** [comment]
Thanks for this. I ran the PR and the main test suite and benchmarks pass.

One remaining edge case: the audience and PKCE checks rely on the file-level MCP marker but do not require the candidate to be inside an MCP handler. Could you please add regression tests for unrelated OAuth flows in an MCP-containing file, or tighten the scope accordingly?

Once that is covered and CI reports pass, this should be ready.

**Comment by SetraTheXX** [comment]
Thanks @asamassekou10 , fixed in 04a7364. Audience and PKCE checks now ignore unrelated OAuth flows in MCP-containing files while preserving genuine MCP and .mcp.json detections. Added regressions across JS, Python, Ruby, and Go; validation and benchmarks still pass with only the existing baseline failures.

**Comment by SetraTheXX** [comment]
I also fixed the remaining JS callback ownership false positive and added regression coverage for sibling tool registrations. The focused OAuth suite is now 67/67, the full suite remains stable with the same five baseline failures, and all corpus benchmarks still pass. No further changes are planned unless CI or review surfaces a concrete issue.

**Comment by asamassekou10** [comment]
Thanks for the contribution. I kept your original work and added a follow-up for registered Ruby and Go MCP handlers, including regression tests. All 584 tests are passing, and this should now be ready to merge.
